File: go/day6.go
package main

import (
	"aoc2024/ds"
	"iter"
	"maps"
	"slices"
)

type day6 struct{}

func (d day6) Part1(lines iter.Seq[string]) any {
	mx := parseMatrix(lines)
	start := findPoints(mx, '^')[0]
	return len(findSeen(mx, start))
}

func (d day6) Part2(lines iter.Seq[string]) any {
	mx := parseMatrix(lines)
	start := findPoints(mx, '^')[0]

	visited := maps.Keys(findSeen(mx, start))

	loops := 0
	for v := range visited {
		if mx[v.x][v.y] == '#' || v.x == start.x && v.y == start.y {
			continue
		}
		mxc := make([][]byte, len(mx))
		for i := range mx {
			mxc[i] = make([]byte, len(mx[i]))
			copy(mxc[i], mx[i])
		}

		mxc[v.x][v.y] = '#'
		if isLoop(mxc, start) {
			loops++
		}
	}

	return loops
}

func findSeen(mx [][]byte, start Point) ds.Set[Point] {
	seen := ds.Set[Point]{}
	seen.Add(start)
	dir := Point{-1, 0}
	n := start
	np := start.Plus(dir)
	for np.isInRage(len(mx)) {
		if mx[np.x][np.y] == '#' {
			np = n
			dir = dir.RotateRight()
		} else {
			seen.Add(np)
			n = np
			np = np.Plus(dir)
		}
	}
	return seen
}

type pdir struct {
	p, dir Point
}

func isLoop(mx [][]byte, start Point) bool {
	// using a bool slice since Set[PointDir] aka map[PointDir]struct{} is too slow
	seen := make([]bool, len(mx)*len(mx)*4)
	dir := Point{-1, 0}
	n := start
	np := start.Plus(dir)
	for np.isInRage(len(mx)) {
		if mx[np.x][np.y] == '#' {
			np = n
			dir = dir.RotateRight()
		} else {
			idx := (np.x*len(mx)+np.y)*4 + slices.Index(Dirs, dir)
			if seen[idx] {
				return true
			}
			seen[idx] = true
			n = np
			np = np.Plus(dir)
		}
	}

	return false
}

File: ocaml/lib/day6.ml
open Point

let count_trues (arr : bool array array) =
  Array.fold_left
    (Array.fold_left (fun acc it -> if it then acc + 1 else acc))
    0 arr

let find_point mx c =
  let i =
    Option.get @@ Array.find_index (fun s -> Array.exists (Char.equal c) s) mx
  in
  let j = Option.get @@ Array.find_index (Char.equal c) mx.(i) in
  Point.create i j

let parse_matrix input =
  Array.of_list input |> Array.map String.to_seq |> Array.map Array.of_seq

let index_of p arr =
  Option.get (Array.find_index (fun e -> e.x = p.x && e.y = p.y) arr)

let find_seen mx p dir =
  let rows, cols = (Array.length mx, Array.length mx.(0)) in
  let seen = Array.make_matrix rows cols false in
  let _ = seen.(p.x).(p.y) <- true in
  let rec aux mx p dir seen =
    let np = Point.add p dir in
    if not (Point.in_range rows cols np) then seen
    else
      match mx.(np.x).(np.y) with
      | '#' -> aux mx p (Point.rotate_right dir) seen
      | _ ->
          seen.(np.x).(np.y) <- true;
          aux mx np dir seen
  in
  aux mx p dir seen

let part1 input =
  let mx = parse_matrix input in
  let start = find_point mx '^' in
  count_trues @@ find_seen mx start { x = -1; y = 0 }

let part2 input =
  let mx = parse_matrix input in
  let rows, cols = (Array.length mx, Array.length mx.(0)) in
  let starting_dir = Point.create (-1) 0 in
  let is_loop mx p dir =
    let seen = Array.make (rows * cols * 4) false in
    let rec aux mx p dir =
      let np = Point.add p dir in
      if not (Point.in_range rows cols np) then false
      else
        match mx.(np.x).(np.y) with
        | '#' -> aux mx p (Point.rotate_right dir)
        | _ ->
            let idx =
              (((np.x * rows) + np.y) * 4) + index_of dir Point.ortogonals
            in
            if seen.(idx) then true
            else
              let () = seen.(idx) <- true in
              aux mx np dir
    in
    aux mx p dir
  in

  let start = find_point mx '^' in
  let seen_part1 = find_seen mx start starting_dir in

  let count = ref 0 in
  for i = 0 to rows - 1 do
    for j = 0 to cols - 1 do
      if seen_part1.(i).(j) then (
        mx.(i).(j) <- '#';
        if is_loop mx start starting_dir then incr count;
        mx.(i).(j) <- '.')
    done
  done;
  !count
